design/Pipelines.md

```diff
@@ -0,0 +1,277 @@
+   Translation from a ```BlendState``` to a ```D3D12_RENDER_TARGET_BLEND_DESC``` is trivial.
+ - **Metal**: the ```BlendStates``` will be used to fill all of the data for a ```MTLRenderPipelineColorAttachmentDescriptor``` but ```pixelFormat```.
+   Translation of individual members is trivial.
+ - **Vulkan**: the ```BlendStates``` will be translated to elements of ```pAttachments``` in the ```VkPipelineColorBlendStateCreateInfo```.
+   Translation of individual members is trivial.
+
+Open question:
+should enablement of independent attachment blend state be explicit like in D3D12 or explicit? Should alpha to coverage be part of the multisample state or the blend state?
+
+### Depth stencil state
+
+Mismatch:
+ - D3D12 doesn’t have per-face stencil read and write masks.
+ - In Metal the depth stencil state is built and bound separately from the pipeline state.
+
+```cpp
+enum CompareFunction {
+    CompareFunctionNever,
+    CompareFunctionLess,
+    CompareFunctionLessEqual,
+    CompareFunctionGreater,
+    CompareFunctionGreaterEqual,
+    CompareFunctionEqual,
+    CompareFunctionNotEqual,
+    CompareFunctionAlways,
+};
+
+enum StencilOperation {
+    StencilOperationKeep,
+    StencilOperationZero,
+    StencilOperationReplace,
+    StencilOperationInvert,
+    StencilOperationIncrementClamp,
+    StencilOperationDecrementClamp,
+    StencilOperationIncrementWrap,
+    StencilOperationDecrementWrap,
+};
+
+struct StencilFaceDescriptor {
+    CompareFunction stencilCompare;
+    StencilOperation stencilPass;
+    StencilOperation stencilFail;
+    StencilOperation depthFail;
+};
+
+struct DepthStencilStateDescriptor {
+    CompareFunction depthCompare;
+    StencilFaceDescriptor front;
+    StencilFaceDescriptor back;
+    int stencilReadMask;
+    Int stencilWriteMask;
+};
+
+DepthStencilState* CreateDepthStencilState(Device* device, DepthStencilDescriptor* descriptor);
+```
+
+Translation to backing APIs would be the following:
+ - **D3D12**: ```DepthStencilState``` translates trivially to a ```D3D12_DEPTH_STENCIL_DESC```.
+   ```DepthEnable``` would be set as ```depthCompare != Always```.
+ - **Metal**: ```DepthStencilState``` translates trivially to ```MTLDepthStencilDescriptor``` except that front and back stencil masks have to be set to the single stencil mask value from WebGPU.
+   When a pipeline is bound, the corresponding depth-stencil state is bound at the same time.
+ - **Vulkan**: ```DepthStencilState``` translates trivially to ```VkPipelineDepthStencilStateCreateInfoxcept``` except that front and back stencil masks have to be set to the single stencil mask value from WebGPU.
+   ```depthTestEnable``` would be set to ```depthCompare != Always```.
+
+Open question:
+what about Vulkan’s ```VkPipelineDepthStencilStateCreateInfo::depthBoundTestEnable```?
```

Comment:
partially supported by d3d12, but I don't see it on Metal: https://msdn.microsoft.com/en-us/library/windows/desktop/mt492556(v=vs.85).aspx